bibo/query.py

```diff
@@ -1,81 +1,96 @@
+import collections
+import collections.abc
 import itertools
 
 import click
 
-from . import internals
+from . import internals, models
 
 
 def search(data, search_terms):
     if isinstance(search_terms, str):
         search_terms = [search_terms]
     search_terms = iter(search_terms)
-    return _recursive_search(internals.bib_entries(data), search_terms)
+    results = (models.SearchResult(e, {}) for e in internals.bib_entries(data))
+    return _recursive_search(results, search_terms)
 
 
-def _recursive_search(data, search_terms):
+def _recursive_search(results, search_terms):
     try:
         search_term = next(search_terms)
-        return _recursive_search(
-            (e for e in data if _is_matching(e, search_term)), search_terms,
-        )
+        # Calculate match with search term and update results
+        results = (_update_result(r, _match(r.entry, search_term)) for r in results)
+        # Drop Nones with empty match
+        results = (r for r in results if r)
+        return _recursive_search(results, search_terms)
     except StopIteration:
-        return data
+        return results
 
 
-def _is_matching(entry, search_term):
+def _match(entry, search_term):
+    """
+    Return a similar structure to an entry (nested dict) with matching strings
+    as values.
+    """
+    if search_term.lower() in entry["key"].lower():
+        return {"key": internals.match_case(search_term, entry["key"])}
```

Comment:
Why end here? I mean, this is a good indication of match, but not a very useful/interesting piece of information potentially compared to what you get from e.g. a `key` _and_ `title` match.

Reply:
You are absolutely right! This is the first check as searching by key can sometime contradict searching by specific field and value. For example, if the key `krappers:2020` appears in the database, you want the search `bibo list krappers:2020` to handle this as a key first, and not as an entry with the field `krappers` equals to 2020.
Anyway, the rest of the match should be filled and highlighted. Will push a fix later.